src/unittest/test_moveaction.cpp

```diff
@@ -0,0 +1,178 @@
+/*
+Minetest
+Copyright (C) 2022 Minetest core developers & community
+
+This program is free software; you can redistribute it and/or modify
+it under the terms of the GNU Lesser General Public License as published by
+the Free Software Foundation; either version 2.1 of the License, or
+(at your option) any later version.
+
+This program is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU Lesser General Public License for more details.
+
+You should have received a copy of the GNU Lesser General Public License along
+with this program; if not, write to the Free Software Foundation, Inc.,
+51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+*/
+
+#include "test.h"
+#include "test_config.h"
+
+#include "mock_inventorymanager.h"
+#include "mock_server.h"
+#include "mock_serveractiveobject.h"
+
+#include <scripting_server.h>
+
+
+class TestMoveAction : public TestBase
+{
+public:
+	TestMoveAction() { TestManager::registerTestModule(this); }
+	const char *getName() { return "TestMoveAction"; }
+
+	void runTests(IGameDef *gamedef);
+
+	void testMove(ServerActiveObject *obj, IGameDef *gamedef);
+	void testMoveSomewhere(ServerActiveObject *obj, IGameDef *gamedef);
+	void testMoveUnallowed(ServerActiveObject *obj, IGameDef *gamedef);
+	void testMovePartial(ServerActiveObject *obj, IGameDef *gamedef);
+	void testSwap(ServerActiveObject *obj, IGameDef *gamedef);
+	void testSwapFromUnallowed(ServerActiveObject *obj, IGameDef *gamedef);
+	void testSwapToUnallowed(ServerActiveObject *obj, IGameDef *gamedef);
+};
+
+static TestMoveAction g_test_instance;
+
+void TestMoveAction::runTests(IGameDef *gamedef)
+{
+	MockServer server;
+
+	ServerScripting server_scripting(&server);
+	server_scripting.loadMod(Server::getBuiltinLuaPath() + DIR_DELIM "init.lua", BUILTIN_MOD_NAME);
+	server_scripting.loadMod(Server::getBuiltinLuaPath() + DIR_DELIM "tests" +
+		DIR_DELIM "helper_moveaction.lua", BUILTIN_MOD_NAME);
+
+	MetricsBackend mb;
+	ServerEnvironment server_env(nullptr, &server_scripting, &server, "", &mb);
```

Comment:
would mocking `ServerEnvironment` not work here?
I'd like to avoid changing the real implementation for the sake of tests if possible.

Reply:
Nope, `ServerActiveObject` accepts only `ServerEnvironment`.

Reply:
It looks like this PR doesn't substantively change `ServerEnvironment` anyway.